src/soc/ibm/power9/tor.c

```diff
@@ -0,0 +1,500 @@
+			       section_size);
+
+		*data_buf_size = section_size;
+	} else {
+		die("Unhandled TOR ring access operation!");
+	}
+}
+
+/* Retrieves an overlay ring in both compressed and uncompressed forms */
+static void get_overlays_ring(struct tor_hdr *overlays_section,
+			      uint16_t ring_id, void *rs4_buf, void *raw_buf)
+{
+	uint32_t uncompressed_bit_size = 0;
+	uint32_t rs4_buf_size = 0xFFFFFFFF;
+
+	tor_access_ring(overlays_section, ring_id, UNDEFINED_PPE_TYPE, 0,
+			rs4_buf, &rs4_buf_size, GET_RING_DATA);
+
+	rs4_decompress(raw_buf, raw_buf + MAX_RING_BUF_SIZE/2,
+		       MAX_RING_BUF_SIZE/2, &uncompressed_bit_size,
+		       (struct ring_hdr *)rs4_buf);
+}
+
+/* Decompress ring, modify it to leave only data allowed by overlay mask and
+ * compress back */
+static void apply_overlays_ring(struct ring_hdr *ring, uint8_t *rs4_buf,
+				const uint8_t *raw_buf)
+{
+	uint8_t *data = rs4_buf;
+	uint8_t *care = rs4_buf + MAX_RING_BUF_SIZE/2;
+	const uint8_t *overlay = raw_buf + MAX_RING_BUF_SIZE/2;
+	uint32_t uncompressed_bit_size;
+
+	rs4_decompress(data, care, MAX_RING_BUF_SIZE/2, &uncompressed_bit_size,
+		       ring);
+
+	/*
+	 * Copies bits from raw_buf into both data and care only if bit at the
+	 * same index in overlay is set.
+	 */
+	for (uint32_t bit = 0; bit < uncompressed_bit_size; ++bit) {
+		const int byte_idx = bit/8;
+		const uint8_t bit_mask = (0x80 >> bit%8);
+		if (overlay[byte_idx] & bit_mask) {
+			if (raw_buf[byte_idx] & bit_mask) {
+				data[byte_idx] |= bit_mask;
+				care[byte_idx] |= bit_mask;
+			} else {
+				data[byte_idx] &= ~bit_mask;
+				care[byte_idx] &= ~bit_mask;
+			}
+		}
+	}
+
+	rs4_compress(ring, MAX_RING_BUF_SIZE, data, care, uncompressed_bit_size,
+		     be32toh(ring->scan_addr), be16toh(ring->ring_id));
+}
+
+static void apply_overlays_to_gptr(struct tor_hdr *overlays_section,
+				   struct ring_hdr *ring, uint8_t *rs4_buf,
+				   uint8_t *raw_buf)
+{
+	get_overlays_ring(overlays_section, be16toh(ring->ring_id), rs4_buf,
+			  raw_buf);
+
+	/* raw_buf is passed from get_overlays_ring(), rs4_buf is just reused */
+	apply_overlays_ring(ring, rs4_buf, raw_buf);
+}
+
+static void tor_append_ring(struct tor_hdr *ring_section,
+			    uint32_t *ring_section_size, uint16_t ring_id,
+			    enum ppe_type ppe_type, uint8_t instance_id,
+			    struct ring_hdr *ring)
+{
+	uint16_t ring_offset;
+	uint32_t ring_size;
+
+	struct ring_put_info put_info;
+	uint32_t put_info_size = sizeof(put_info);
+
+	tor_access_ring(ring_section, ring_id, ppe_type, instance_id, &put_info,
+			&put_info_size, GET_RING_PUT_INFO);
+
+	if (*ring_section_size - put_info.chiplet_offset > MAX_TOR_RING_OFFSET)
+		die("Not enough space to append ring to TOR section!");
+
+	ring_offset = htobe16(*ring_section_size - put_info.chiplet_offset);
+	ring_size = be16toh(ring->size);
+
+	memcpy(ring_section + put_info.ring_offset, &ring_offset, ring_offset);
+	memcpy(ring_section + *ring_section_size, ring, ring_size);
```

Comment:
These look suspicious. I assume `put_info.ring_offset` and `ring_section_size` are in bytes, and not in multiples of `tor_hdr` sizes. Also, first `memcpy` copies a lot of data from stack and even above stack.

```suggestion
	memcpy((void*)ring_section + put_info.ring_offset, &ring_offset, sizeof(ring_offset));
	memcpy((void*)ring_section + *ring_section_size, ring, ring_size);
```